README.md
# agent-migration

#### Table of Contents

1. [Overview](#overview)
2. [Description](#module-description)
3. [Setup](#setup)
* [What agent-migration affects](#what-[agent-migration]-affects)
* [Setup requirements](#setup-requirements)
* [Beginning with this module](#beginning-with-agent-migration)
4. [Usage](#usage)
5. [Reference](#reference)
5. [Limitations - OS compatibility, etc.](#limitations)
6. [Development - Guide for contributing to the module](#development)

## Overview

`agent-migration` helps in migrating puppet agents to a new Puppet master

## Description

This module contains puppet manifests to install the mcollective agent files.
It contains a simple rpc agent and an mcollective subcommand (ie: application)
## Setup

`include agent-migration`

### Setup requirements

* Puppet Enterprise

## Beginning with this module

### Usage

#### High level process to use this module
* Ensure the "to be" master does not have a cert signed for the nodes you are migrating.
* Test your mcollective node filters
* Test one run of the `migrate` command on a single node
* Confirm the cert request appeared on the new master
    * this takes ~2 minutes.
* Slowly progress through the rest of your nodes in small batches, always checking that certs have been requested.

To follow that process, read through the warnings and command help in the Usage section below.


#### RPC Commands Available

Supplementary command documentation to that on the `plugin doc` command documentation.

* `agent_from_3_to_4`
    * use this if you want to upgrade an agent node from a puppet 3 master to a puppet 4 master.
    * this command is asynchronous
    * this command uninstalls puppet, and curls the version of puppet from the new master.  It is **important** to test one node at a time, and do this in a controlled manner, as if your node fails, you will have to intervene manually.
    * Example: `mco rpc migrate agent_from_3_to_4 to_fqdn=newmaster.domain.com to_ip=10.1.1.10 -v -I /filter/`
* `puppet_agent`
    * once you are on puppet 4, you can use this command.
    * migrates a puppet 4 agent to another puppet 4 agent.
    * Example: `mco rpc migrate puppet_agent to_fqdn=newmaster.domain.com to_ip=10.1.1.10 -v -I /filter/`
* `test_activation`
    * `mco find <filter>` is useful to test a query, but rpc agents can still refuse to activate.
    * if this agent isn't active on a node, this command will tell you.
    * use this in conjunction with your mcollective node filter for a definitive test before running * Example:  `mco rpc migrate test_activation -v -I /filter/`

#### `migrate` mco subcommand

***This is the recommended way to use this mco agent***

`mco help migrate` will show you current usage documentation.

This command **requires puppet 4** as it wraps around the `puppet_agent` rpc plugin.

An example of how to run the migrate subcommand.
`mco migrate agent <options>`

### Learn MCollective Basics

Familiarize yourself with running mCollective on the command line:
https://docs.puppetlabs.com/mcollective/reference/basic/basic_cli_usage.html

You must run mCollective from a user authorized to do so.  On PE, out of the box, the only user that has this
capability is the peadmin user.   All commands should be run as that user.

Ex: `runuser -l peadmin -c 'mco <subcommand>'`.   This document assumes you will do this for all mco commands.

### View documentation on the command line

Familiarize yourself with the documentation of this module"
`runuser -l peadmin -c 'mco plugin doc migrate'`

Example output of `plugin doc` for this plugin:

```
ACTIONS:
========
   agent, agent_from_3_to_4

   agent action:
   -------------
          - to be implemented -
   agent_from_3_to_4 action:
   -------------------------
       Uninstalls puppet, reinstalls from designated master curl script

       INPUT:
           to_fqdn:
              Description: Master FQDN to direct the agent to
                   Prompt: new master fqdn
                     Type: string
                 Optional: false
               Validation: .*
                   Length: 1024

           to_ip:
              Description: Master IP Address to direct the agent to
                   Prompt: new master ip
                     Type: string
                 Optional: false
               Validation: .*
                   Length: 1024


       OUTPUT:
           msg:
              Description: error output if it failed
               Display As: errors or messages from client
            Default Value: unknown
```

### Warning
The subcommand `agent_from_3_to_4` will uninstall puppet and reinstall puppet from the master you supply. This will fail if any of the following conditions exist:

* the new master isn't available
* there is an existing cert on the new master.
    * to ensure there isn't: run `puppet node clean <client_cert_name_here>` on the new master.

If the migration fails, you will have to log into the machine you tried to migrate.  In other words, be careful.

As the `agent_from_3_to_4` command is asynchronous, it is helpful to know the reinstall typically takes 2-3 minutes to complete.


### Executing commands

As always with mcollective, **first test your node filter**.  You probably don't want to run this command against every agent!

This command: `mco rpc migrate test_activation -I /hostname_regex/` will tell you what nodes would run on an activate command.
* if a node doesn't appear in the list and you expect it to,  the plugin may not be deployed.
    * ensure the module is included in this nodes profile.
    * ensure puppet has run to place the plugin on the node.

*Note: these commands will filter out the puppet master itself, as migrating the masters agent to another master would be bad.   Therefore, when testing your filter - you can ignore the puppet master node if it shows up.*

Take the filter above (`-I /hostname_regex/`) and use it in running the migration:

`mco rpc migrate agent_from_3_to_4 to_fqdn=devcorepptl918.matrix.sjrb.ad to_ip=10.15.185.90  -I /hostname_regex/ -v`

Be sure to add the -v to get verbose results.

#### Summary of commands you may wish to use
If your new master is devcorepptl918.matrix.sjrb.ad, at ip 10.15.185.90, and you are migrating a vm called vm_001:

1. `sudo runuser -l peadmin -c 'mco rpc migrate test_activation -I /vm_001/'`
2. `sudo runuser -l peadmin -c 'mco rpc migrate agent_from_3_to_4 to_fqdn=devcorepptl918.matrix.sjrb.ad to_ip=10.15.185.90  -I /vm_001/ -v'`

## Limitations

* Puppet Enteprise
* Tested On:
    * RHEL 6.4
    * CentOS 6.4
* the `agent_from_3_to_4` currently has a hard coded url to a puppet tar. It won't work outside our firewall .. yet.
    * TODO: allow injecting url for tarball to uninstall puppet 3.

## Development

Pull Requests are welcome.  Please rebase and squash before submitting, and use a feature branch.

For more details: CONTRIBUTING.md.

## Release Notes/Contributors/Etc

See the `CHANGELOG`.


CHANGELOG
CHANGELOG
=========

next
____
 - removed dependency on the `to_ip` option
 - windows support

0.0.5
____

## Notes
 - adding support for a simpler migration agent between puppet 4 masters.
 - added the `mco migrate` subcommand that wraps the `mco rpc migrate puppet_agent` plugin.

0.0.4
____

 - prevented module from failling if included on puppet open source


0.0.3
____
 - initial release


manual_reinstall_on_puppet3.sh
#!/bin/bash

#change these as required
NEW_PUPPET_MASTER='devcorepptl900.matrix.sjrb.ad'
NEW_PUPPET_MASTER_IP='10.15.184.156'
PUPPET_3_ENVIRONMENT='feature_master_migrator'

cd /tmp

if [[ ! -f /tmp/puppet-enterprise-2015.2.3-el-6-x86_64.tar.gz ]]; then
   /usr/bin/wget --no-check-certificate https://prd-artifactory.sjrb.ad:8443/artifactory/shaw-private-core-devops-ext-release/com/puppetlabs/puppet-enterprise/2015.2.3/puppet-enterprise-2015.2.3-el-6-x86_64.tar.gz
   /bin/tar -xvf puppet-enterprise-2015.2.3-el-6-x86_64.tar.gz
fi
/tmp/puppet-enterprise-2015.2.3-el-6-x86_64/puppet-enterprise-uninstaller -pdy
rm -rf /tmp/puppet-enter*
rm -rf /tmp/uninstall*
rm -rf /tmp/tmp.*

rm -rf /etc/yum.repos.d/pe*
rm -rf /etc/yum.repos.d/puppet*

sed -i '/puppet/c\$NEW_PUPPET_MASTER_IP puppet $NEW_PUPPET_MASTER' /etc/hosts

curl -k https://$NEW_PUPPET_MASTER:8140/packages/current/install.bash | sudo bash

echo "environment = $PUPPET_3_ENVIRONMENT" >> /etc/puppetlabs/puppet/puppet.conf

rm -rf /etc/puppetlabs/puppet/ssl/*

echo "certs have been cleaned here.. clean them on the master.. you have 10 seconds"

sleep 10s

echo "running puppet ! ! ! ! ! ! ! !  go sign the cert!"

puppet agent -t --waitforcert 10
puppet agent -t

echo "ran puppet 2 times on clean install"


files/agent/migrate.rb
module MCollective
  module Agent
    class Migrate<RPC::Agent
      require 'date'

      action 'puppet_reinstall' do
        to_fqdn = request[:to_fqdn]
        run_reinstall_migration(to_fqdn)
      end

      action 'puppet_agent' do
        to_fqdn = request[:to_fqdn]
        run_migration(to_fqdn)
      end

      action 'test_activation' do
        reply[:fqdn] = run("hostname -f", :stdout => :out, :stderr => :err)
      end

      def run_migration(to_fqdn)
        Log.info("Executing:   #{cmd}")
        reply[:exitstatus] = "initialized"
        reply[:command_list] = []
        kernel=run("facter kernel", :stdout => kernel)
        case kernel
        when "Windows"
          run_migration_windows(to_fqdn)
        when "Linux"
          run_migration_linux(to_fqdn)
        else
          reply[:msg] = "kernel #{kernel} not supported"
          exit
        end
      end

      def run_migration_windows(to_fqdn)
        update_puppet_conf_and_nuke_ssl_cmd="powershell -file c:/windows/temp/migrate_to_new_master.ps1 -to_fqdn #{to_fqdn}"
        run_command(update_puppet_conf_and_nuke_ssl_cmd)

        certlink="https://#{to_fqdn}/#/node_groups/inventory/nodes/certificates"
        reply[:msg] = "Migration complete. Go look for your cert at #{certlink}"
        reply[:certlink]=certlink
      end

      def run_migration_linux(to_fqdn)
        # update_puppet_conf_cmd="puppet resource ini_setting 'server' ensure=present path='/etc/puppetlabs/puppet/puppet.conf' section='main' setting='server' value='#{to_fqdn}'" #hmm, doesn't work?
        update_puppet_conf_cmd="sed -i '/server/cserver = #{to_fqdn}' /etc/puppetlabs/puppet/puppet.conf"
        nuke_ssl_dir_cmd="rm -rf /etc/puppetlabs/puppet/ssl"
        # nuke_ssl_dir_cmd="puppet resource file '/etc/puppetlabs/puppet/ssl' ensure=absent force=true"   #hmm, doesn't work?
        restart_agent_cmd='/etc/init.d/puppet restart'
        run_command(update_puppet_conf_cmd)
        run_command(nuke_ssl_dir_cmd)
        run_command(restart_agent_cmd)

        certlink="https://#{to_fqdn}/#/node_groups/inventory/nodes/certificates"
        reply[:msg] = "Migration complete. Go look for your cert at #{certlink}"
        reply[:certlink]=certlink
      end

      def run_reinstall_migration(to_fqdn)
        reinstall_script=''
          reinstall_script = <<REINSTALL
          /usr/bin/wget --no-check-certificate https://prd-artifactory.sjrb.ad:8443/artifactory/shaw-private-core-devops-ext-release/com/puppetlabs/puppet-enterprise/2015.2.3/puppet-enterprise-2015.2.3-el-6-x86_64.tar.gz
          /bin/tar -xvf puppet-enterprise-2015.2.3-el-6-x86_64.tar.gz
          /bin/bash puppet-enterprise-2015.2.3-el-6-x86_64/puppet-enterprise-uninstaller -pdy
          /bin/rm -rf puppet-enterprise-*
REINSTALL

        full_script = <<OEF
        #!/bin/bash
        set -x
        exec 2> >(logger)
        [ -f /etc/init.d/puppet ] &&  puppet resource service puppet ensure=stopped
        #{reinstall_script}
        /usr/bin/curl -o /tmp/install_puppet.sh -k https://#{to_fqdn}:8140/packages/current/install.bash
        /bin/chmod +x /tmp/install_puppet.sh
        /bin/bash /tmp/install_puppet.sh
        /bin/rm -rf /etc/yum.repos.d/pe_repo.repo
        /bin/rm -rf /etc/yum.repos.d/pc_repo.repo
        /bin/rm -rf /tmp/install_puppet.sh
        [ -f /etc/init.d/puppet ] &&  /usr/local/bin/puppet resource service puppet ensure=running
OEF

        reinstall_file='/tmp/reinstall_puppet_from_new_master.sh'
        File.write(reinstall_file, full_script)

        reply[:msg] = run("nohup /bin/bash #{reinstall_file} &", :stdout => :out, :stderr => :err, :cwd => "/tmp")
        certlink="https://#{to_fqdn}/#/node_groups/inventory/nodes/certificates"
        reply[:msg] = "Migration was triggered, and mcollective will uninstall now. Go look for your cert at #{certlink}"
        reply[:certlink]=certlink
      end

      activate_when do
        #deactivate if any puppet master services exist. Migrating a master would be bad.
        #TODO: use a fact?
        if (File.exists?("/etc/init.d/pe-puppetserver") ||
            File.exists?("/etc/init.d/pe-httpd") ||
            File.exists?("/etc/init.d/pe-console-services") ||
            File.exists?("/etc/init.d/pe-memcached") ||
            File.exists?("/etc/init.d/pe-postgresql") ||
            File.exists?("/etc/init.d/pe-puppet-dashboard-workers") ||
            File.exists?("/etc/init.d/pe-puppetdb")) then
            return false
        else
          return true
        end
      end

      private

      def run_command(cmd)
        Log.info("Executing:   #{cmd}")
        #create reply[:out] and reply[:error] entries in the hash.
        status = run(cmd, :stdout => :out, :stderr => :error)
        reply[:exitstatus] = "Success" if status
        reply[:command_list] << "Command: #{cmd} exited with: #{status}"

        Log.info("Status is:   #{reply[:exitstatus]}")
        Log.info("STDOUT:")
        Log.info(reply[:out])
        Log.info("Execution Success for: #{cmd}")
      end
    end
  end
end
